### ISSUE TITLE
Proposed list of algorithms / package layout

### ISSUE THREAD

**Issue by akonneker** [image-rs/imageproc#9]
Hello,

I'm working on some computer vision problems, and I plan on switching my current matlab prototypes over to rust in the coming months. As such, I would love to make use of the full-fledged image processing crate that doesn't exist yet...

I realize you're in the early stages, but as soon as you have a proposed api and a list of functions/algorithms you'd like implemented, I'd be happy to help. There will be things that I'll probably write myself sooner rather than later which could have a home here., e.g. an implementation of Otsu's method for thresholding.

Also, what is the proposed scope of the library? I'm interested in frequency-space filters, image features like SIFT/SURF, image registration, edge detection, etc.

**Comment by theotherphil** [comment]
Great, I'd be happy for any contributions - there's an awful lot to be written! Otsu thresholding would be a nice addition. 

I don't have a concrete plan for what's going in the library yet, but I was hoping to eventually include everything you mentioned, except probably image registration. My intended split is for image processing/image features to go in this crate and vision/learning stuff in a separate crate. For example I think that corner detection should be here, but RANSAC elsewhere. There will obviously be edge cases, but the line seems reasonably clear.

I plan on adding FAST12 corner detection and HOG features pretty soon. Gradient images are already doable using either image::filter3x3 or imageproc::filter::separable_filter. We'll need to sort out the overlap between image::sample and imageproc::filter at some point. I was planning on adding Canny edge detection soon, but not until I've added a few more basic image processing functions.

As I'm totally new to rust, and also not an expert on image processing or computer vision I'd be very happy for suggestions on APIs, implementations, or features to prioritise.

**Comment by theotherphil** [comment]
I've just created enhancement issues for Otsu and Canny implementations. Please add anything else you'd like or are planning on doing and we can then assign them to contributors and avoid that awkward moment where two people want to commit their own versions of the same algorithm!

**Comment by AaronViviano** [comment]
"My intended split is for image processing/image features to go in this crate and vision/learning stuff in a separate crate."
So I have done professional development on a vision system. There are basically 5 categories of algorithms that the AIA teaches to be relevant for a computer vision system:
*Image Enhancement
*Presence Checking
*Feature Location
*Feature Measurement
*Identification (Barcodes/OCR)

To me it sounds like you'd like this library to simply consist of Image Enhancement while the other algorithms would exists in a vision crate?

**Comment by theotherphil** [comment]
There are already functions for corner detection (FAST), extracting feature descriptors (Haar and HOG), and connected component labelling, so the scope's definitely broader than just image enhancement.

The split I imagined doesn't really fit into the taxonomy you gave. I imagined that there'd be functions helpful for each of those categories, but only the lower level parts of the algorithms.

e.g. 

OCR:
* Yes: functions from image to radial basis coefficients
* No: classifiers for the extracted descriptors
Object recognition: 
* Yes: corner point detection, edge detection, SURF/HOG/SIFT/HAAR/LBP features
* No: boosting, neural nets, SVMs, ...
Image registration: 
* Yes: interest point descriptors, simple patch discrepancy measures
* No: RANSAC/PROSAC, energy minimisation algorithms, ...

I think for now we should just err on the side of including things, and worry more about scope when the higher level stuff's actually been written. We can always split the library apart later. The important thing is just to have somewhere visible for people to see what's been done already and coordinate future work. Ideally we'd eventually end up with a nice ecosystem of pure image processing stuff here, feature extraction code, graph processing algorithms, learning algorithms, and applications, probably split across many crates. But that's a long, long way off. I'm also not sure if the PistonDevelopers powers that be would be happy even in principle for us to start adding a load of computer vision stuff here, but again this isn't a problem until some more code's been written!

**Comment by AaronViviano** [comment]
Hello @theotherphil,
I don't think I can agree with the High/Low level split like you've described. Some of what you've mentioned including wouldn't be what I would consider to be image processing (Image Enhancement) and would be more along the lines of image analysis (the other 4 categories that I've mentioned). I do agree that it wouldn't be a good idea to include any of your no examples. However, I don't think Identification techniques, like OCR, in general belong in a general purpose image processing library.

In addition I can't agree with splitting the library apart latter. I don't think we can expect a user of the library to know that the libraries been split apart if they update their code. It would be different if we could package the libraries together. Regardless I don't think I can spend time working on this project unless there is an upfront plan as to its scope. I'd be happy to look at working on this project if such a plan is developed.

Have a good day and good luck with the library,
Aaron

**Comment by theotherphil** [comment]
Feel free to write one, I'm not that bothered about the nature of the split. If you do this you should get some feedback from the PistonDevelopers admins about whether it's ok to add all the new projects here or if we should create a separate computer vision organisation (or whatever the GitHub term is) instead.

**Comment by bvssvni** [comment]
@theotherphil You are free to push as much as code as you like under PistonDevelopers. Use this repo, to create new ones if you need it. If you later want to move it to a separate computer vision organization, I'm 100% ok with that. @mitchmindtree did this with RustAudio.

@AaronViviano I think you're absolutely right about making a plan upfront. Computer vision is a large field and it would be a mistake to not learn from previous work.

**Comment by theotherphil** [comment]
Thanks, that's good to know.

**Comment by AaronViviano** [comment]
Sure thing @theotherphil . Give me a bit of time to get a plan together. And thank you for the update and information @bvssvni. Its appreciated.

**Comment by AaronViviano** [comment]
Hello @theotherphil. Looks like your actively working on the project and I'm glad to see that. Rust deserves and Imag Proc lib. So I finally remembered to note that I won't be contributing to this project at this time. I've decided to work on other things in my free time, but I didn't want my previous comment to go unanswered. Good luck!
